clang/lib/Sema/SemaBounds.cpp

```diff
@@ -2123,6 +2123,71 @@ namespace {
       }
    }
 
+   void ResetKilledBounds(StmtDeclSetTy &KilledBounds,
+                          BoundsContextTy &BoundsCtx,
+                          Stmt *S) {
+     auto I = KilledBounds.find(S);
+     if (I == KilledBounds.end())
+       return;
+
+     for (const VarDecl *V : I->second) {
+       VarDecl *VD = const_cast<VarDecl *>(V);
+       BoundsExpr *Bounds = VD->getBoundsExpr();
+       assert(Bounds && "Invalid bounds");
+
+       if (isa<CountBoundsExpr>(Bounds))
+         Bounds = ExpandToRange(VD, Bounds);
+       BoundsCtx[VD] = Bounds;
+     }
+   }
+
+   void UpdateBoundsInContext(BoundsMapTy &WidenedBounds,
+                              BoundsContextTy &BoundsCtx) {
+     for (const auto item : WidenedBounds) {
+       VarDecl *V = const_cast<VarDecl *>(item.first);
+       unsigned Offset = item.second;
+
+       BoundsExpr *Bounds = V->getBoundsExpr();
+       if (!Bounds)
+         continue;
+
+       // If the declared bounds are CountBoundsExpr, normalize them to
```

Comment:
After merging #827, all bounds in the bounds context should already be a normalized RangeBoundsExpr.